README
Net-OpenSSH-Mux version 0.01
============================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  blah blah blah

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2011 by Salvador Fandino

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.12.2 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Net::OpenSSH::Mux.

0.01  Wed Mar  2 13:45:01 2011
	- original version; created by h2xs 1.23 with options
		-AXn Net::OpenSSH::Mux



lib/Net/OpenSSH/Mux/Packer.pm
package Net::OpenSSH::Mux::Packer;

use 5.010;

our $VERSION = '0.01';

use strict;
use warnings;
use Carp;

use Net::OpenSSH::Mux::Constants qw(:msg);

require Exporter;
our @ISA = qw(Exporter);
our @EXPORT = do { no strict; grep /^(?:un)?pack_/, keys %{__PACKAGE__ . '::'}};

sub _pack_string { $_[0] .= pack(N => length $_[1]) . $_[1] }

sub _pack_start {
    my $start = length $_[0];
    $_[0] .= "\x00\x00\x00\x00";
    $start;
}

sub _pack_end {
    my $start = $_[1];
    my $end = length $_[0];
    substr($_[0], $start, 4, pack(N => $end - $start - 4));
}

sub pack_msg_hello_msg {
    my $start = _pack_start($_[0]);
    $_[0] .= "\x00\x00\x00\x01\x00\x00\x00\x04"; # $_[0] .= pack(NN => MUX_MSG_HELLO, 4);
    _pack_string($_[0], $_) for (@_[1..$#_]);
    _pack_end($_[0], $start);
}

sub pack_c_alive_check_msg {
    my $start = _pack_start($_[0]);
    $_[0] .= pack NN => MUX_C_ALIVE_CHECK, $_[1];
    _pack_end($_[0], $start)
}

sub pack_c_new_session_msg {
    my (undef, $rid, $cmd, %opts) = @_;
    my $flags = 0;
    my $want_tty = delete $opts{want_tty} || 0;
    my $want_X11_forwarding = delete $opts{want_X11_forwarding} || 0;
    my $want_agent = delete $opts{want_agent} || 0;
    my $subsystem = delete $opts{subsystem} || 0;
    my $escape = delete $opts{escape_char} // 0xffffffff;
    my $term = delete $opts{term} // $ENV{TERM} // 'vt100';
    my $env = delete $opts{env} // [];

    my $start = _pack_start($_[0]);
    $_[0] .= pack('N*' => MUX_C_NEW_SESSION, $rid, 0,
                  $want_tty, $want_X11_forwarding, $want_agent,
                  $subsystem, $escape);
    _pack_string($_[0], $term);
    _pack_string($_[0], $cmd);
    _pack_string($_[0], $_) for @$env;
    _pack_end($_[0], $start);
}

sub pack_c_new_stdio_fwd {
    my (undef, $rid, $connect_host, $connect_port) = @_;
    my $start = _pack_start($_[0]);
    $_[0] .= pack NN => MUX_C_NEW_STDIO_FWD, $rid;
    _pack_string($_[0], $_) for ('', $connect_host, $connect_port);
    _pack_end($_[0], $start);
}

sub pack_c_terminate_msg {
    my $start = _pack_start($_[0]);
    $_[0] .= pack(NN => MUX_C_TERMINATE, $_[1]);
    _pack_end($_[0], $start);
}

sub pack_c_open_fwd {
    my (undef, $rid, $type, $listen_host, $listen_port, $connect_host, $connect_port) = @_;
    my $start = _pack_start($_[0]);
    $_[0] .= pack(NNN => MUX_C_OPEN_FWD, $rid, $type);
    $_[0] .= _pack_string($_) for ($listen_host, $listen_port, $connect_host, $connect_port);
    _pack_end($_[0], $start);
}

sub _unpack_string {
    length $_[0] >= 4 or return ();
    my $len = unpack N => substr($_[0], 0, 4, '');
    return substr($_[0], 0, $len, '');
}

sub unpack_msg_hello {
    length $_[0] >= 4 or return ();
    my $ver = unpack N => substr($_[0], 0, 4);
    my @ext;
    while (length $_[0]) {
        my $ext = _unpack_string($_[0]) // last;
        push @ext, $ext;
    }
    return $ver, @ext;
}

sub unpack_s_alive_msg {
    length $_[0] >= 8 or return ();
    unpack NN => substr($_[0], 0, 8, ''); # rid, pid
}

sub unpack_s_ok_msg {
    length $_[0] >= 4 or return ();
    unpack N => substr($_[0], 0, 4, '');
}

sub unpack_s_error_msg {
    length $_[0] >= 4 or return ();
    my $rid = unpack N => substr($_[0], 0, 4, '');
    my $reason = _unpack_string($_[0]);
    return ($rid, $reason);
}

sub unpack_s_remote_port {
    length $_[0] >= 12 or return ();
    unpack NNN => substr($_[0], 0, 12, '');
}

sub unpack_s_session_opened_msg {
    length $_[0] >= 8 or return ();
    unpack NN => substr($_[0], 0, 8, ''); # rid, sid.
}

sub unpack_s_exit_message_msg {
    length $_[0] >= 8 or return ();
    unpack NN => substr($_[0], 0, 8, ''); # sid, exit code
}

1;


lib/Net/OpenSSH/Mux/Constants.pm
package Net::OpenSSH::Mux::Constants;

use strict;
use warnings;

require Exporter;
our @ISA = qw(Exporter);

use constant MUX_MSG_HELLO           => 0x00000001;
use constant MUX_C_NEW_SESSION       => 0x10000002;
use constant MUX_C_ALIVE_CHECK       => 0x10000004;
use constant MUX_C_TERMINATE         => 0x10000005;
use constant MUX_C_OPEN_FWD          => 0x10000006;
use constant MUX_C_CLOSE_FWD         => 0x10000007;
use constant MUX_C_NEW_STDIO_FWD     => 0x10000008;
use constant MUX_S_OK                => 0x80000001;
use constant MUX_S_PERMISSION_DENIED => 0x80000002;
use constant MUX_S_FAILURE           => 0x80000003;
use constant MUX_S_EXIT_MESSAGE      => 0x80000004;
use constant MUX_S_ALIVE             => 0x80000005;
use constant MUX_S_SESSION_OPENED    => 0x80000006;
use constant MUX_S_REMOTE_PORT       => 0x80000007;

use constant MUX_FWD_LOCAL   => 1;
use constant MUX_FWD_REMOTE  => 2;
use constant MUX_FWD_DYNAMIC => 3;

our %EXPORT_TAGS;
$EXPORT_TAGS{msg} = [do { no strict; grep /^MUX_(?:C|S|MSG)_/, keys %{__PACKAGE__ ."::"} }];
$EXPORT_TAGS{fwd} = [do { no strict; grep /^MUX_FWD_/,         keys %{__PACKAGE__ ."::"} }];

our @EXPORT_OK = map @$_, values %EXPORT_TAGS;

$EXPORT_TAGS{all} = [@EXPORT_OK];

1;


lib/Net/OpenSSH/Mux.pm
package Net::OpenSSH::Mux;

use 5.010;

our $VERSION = '0.01';

use strict;
use warnings;
use Carp;

use Net::OpenSSH::Mux::Packer;
use Net::OpenSSH::Mux::Constants qw(:all);

use Socket::PassAccessRights;
use Errno ();

our $debug;

sub _debug { print STDERR '# ', (map { defined($_) ? $_ : '<undef>' } @_), "\n" }

sub _hexdump {
    no warnings qw(uninitialized);
    my $data = shift;
    while ($data =~ /(.{1,32})/smg) {
        my $line=$1;
        my @c= (( map { sprintf "%02x",$_ } unpack('C*', $line)),
                (("  ") x 32))[0..31];
        $line=~s/(.)/ my $c=$1; unpack("c",$c)>=32 ? $c : '.' /egms;
        print STDERR "#> ", join(" ", @c, '|', $line), "\n";
    }
}

sub new {
    my ($class, %opts) = @_;
    my $socket = delete $opts{socket} // croak "missing argument 'socket'";
    my $self = { _socket => $socket,
                 _error => '',
                 _rid => 1 };
    bless $self, $class;
    $self->hello;
    $self;
}

sub _next_rid { shift->{_rid}++ }

sub _set_error {
    my $self = shift;
    $self->{_error} = join(': ', @_);
    $debug and _debug "error set to $self->{_error}"
}

sub _send_msg {
    my ($self) = @_;
    my $len = length $_[1];
    my $off = 0;
    my $socket = $self->{_socket};

    if ($debug) {
        _debug "_send_msg(socket: $socket, buffer len: $len)";
        _hexdump $_[1];
    }

    while ($off < $len) {
        my $bytes = syswrite($socket, $_[1], 16 * 1024, $off);
        if ($bytes) {
            $off += $bytes;
        }
        else {
            $self->_set_error("connection to master lost: $!");
            return undef;
        }
    }
    for my $fh (@_[2..$#_]) {
        unless (Socket::PassAccessRights::sendfd(fileno($socket), fileno($fh))) {
            redo if ($! == Errno::EAGAIN() or $! == Errno::EINTR());
            $self->_set_error("unless to sendfd: $!");
            return undef;
        }
    }
    return 1;
}

sub _recv_data {
    my ($fd, $len) = @_;
    $_[2] = '';
    while (1) {
        my $off = length($_[2]);
        if ($off >= $len) {
            if ($debug) {
                _debug "recv_data($fd, $len) => ", length($_[2]), " bytes read";
                _hexdump($_[2]);
            }
            return 1
        }
        my $bytes = sysread($fd, $_[2], $len - $off, $off);
        unless ($bytes) {
            redo if ($! == Errno::EAGAIN() or $! == Errno::EINTR());
            $debug and _debug "sysread failed: $!";
            return undef;
        }
    }
}

sub _recv_msg {
    my ($self) = @_;
    my $socket = $self->{_socket};
    if (_recv_data($socket, 8, $_[1])) {
        my ($size, $cmd) = unpack NN => $_[1];
        if (_recv_data($socket, $size - 4, $_[1])) {
            $debug and _debug "received message of $size bytes, $cmd = $cmd";
            return $cmd;
        }
    }
    $self->_set_error("connection to master lost");
    return undef;
}

sub hello {
    my $self = shift;
    my $buf = '';
    pack_msg_hello_msg($buf, 'keep-alive' => 1);
    $self->_send_msg($buf);
    my $cmd = $self->_recv_msg($buf);
    unless ($cmd == MUX_MSG_HELLO) {
        $self->_set_error("response missmatch, expecting " . MUX_MSG_HELLO . " cmd, got $cmd");
        return undef;
    }
    my @r = unpack_msg_hello($buf);
    unless (@r) {
        $self->_set_error("corrupted msg_hello packed");
        return undef;
    }
    return (wantarray ? @r : $r[0]);
}

sub _send_c_new_session {
    my ($self, $cmd, $opts, $fdin, $fdout, $fderr) = @_;
    $opts //= {};
    my $rid = $self->_next_rid;
    my $buf = '';
    pack_c_new_session_msg($buf, $rid, $cmd, %$opts);
    $self->_send_msg($buf, $fdin, $fdout, $fderr) or return ();
    return $rid;
}

sub _c_new_session {
    my $self = shift;
    my $rid = $self->_send_c_new_session(@_) // return ();
    my $cmd = $self->_recv_msg(my $buf);
    given ($cmd) {
        when(MUX_S_SESSION_OPENED) {
            my ($rid1, $sid) = unpack_s_session_opened_msg($buf);
            unless ($rid == $rid1) {
                $self->_set_error("query/reply id mismatch");
                return ();
            }
            return $sid;
        }
        when ([MUX_S_PERMISSION_DENIED, MUX_S_FAILURE]) {
            my ($rid1, $reason) = unpack_s_error_msg($buf);
            unless ($rid == $rid1) {
                $self->_set_error("query/reply id mismatch");
                return ();
            }
            $self->_set_error("unable to open new session: ", $reason);
            return ();
        }
        default {
            $self->_set_error("unexpected response $cmd");
            return ();
        }
    }
}

sub _wait_for_exit_message {
    my ($self, $sid) = @_;
    my $cmd = $self->_recv_msg(my $buf) // return ();
    if ($cmd != MUX_S_EXIT_MESSAGE) {
        $self->_set_error("unexpected response $cmd");
        return ();
    }
    my ($sid1, $exit) = unpack_s_exit_message_msg($buf);
    unless ($sid == $sid1) {
        $self->_set_error("session id mismatch, received $sid1, expected $sid");
        return ();
    }
    return $exit;
}

sub system {
    my ($self, $cmd) = @_;
    my $sid = $self->_c_new_session($cmd, undef, \*STDIN, \*STDOUT, \*STDERR) // return ();
    return $self->_wait_for_exit_message($sid)
}

1;
__END__

=head1 NAME

Net::OpenSSH::Mux - OpenSSH Mux client

=head1 SYNOPSIS

  use Net::OpenSSH::Mux;

=head1 DESCRIPTION

Stub documentation for Net::OpenSSH::Mux, created by h2xs. It looks like the
author of the extension was negligent enough to leave the stub
unedited.

=head2 EXPORT

None by default.



=head1 SEE ALSO

Mention other useful documentation such as the documentation of
related modules or operating system documentation (such as man pages
in UNIX), or any relevant external documentation such as RFCs or
standards.

If you have a mailing list set up for your module, mention it here.

If you have a web site set up for your module, mention it here.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by Salvador Fandino E<lt>sfandino@yahoo.comE<gt>.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.12.2 or,
at your option, any later version of Perl 5 you may have available.

=cut
